---
layout: post
title:  "Putting the SCE in... 'Sceince': A Rushed Introduction to Single Cell Experiments"
date: 2024-10-02
description:  Since I'm a Statistcian^TM^, I am legally unable to spell. Anyways, this is a brief introduction to using and analyzing SCE objects in R.   
image: "/assets/img/image5.jpg"
display_image: false  # change this to true to display the image below the banner 
---
<p class="intro"><span class="dropcap">S</span>o, you woke up this morning and wondered "what in the heck is an SCE object?". Just kidding, no one does that. That's why I'm here to tell you about something most people will never use!</p>

## The Boring Stuff
### What is the SCE object?

I'll start with the basics. SCE stands for 'Single-Cell Experiment'. So an SCE object, you guessed it, contains biological data for cells. "Why would you want to work with biological data?" you ask. "Because I'm a masochist," I answer. No, I'm a biostatistican, but same difference. At least I'm not an actuary. Those guys really *are* masochists. (I'm joking, please don't hurt me!)

The main assay, which is essentially a data frame, is made up of thousands of rows representing certain genes and columns representing certain (biological) cells. The data I'm playing with here has 6,260 gene expressions (rows) and 65,246 individual cells (columns), meaning I have a grand total of a whopping *408,439,960* (table) cells. Yeah, it's a lot. In fact, it has to use a sparse matrix - a matrix that collapses all cells with '0' into commas - to save space. Even then, this particular object is about 1.3 gigabytes.

### Where is the SCE object?

BiocManager is your best friend. Here's how you'd install it:

{%- highlight python -%}
install.packages("BiocManager", repos = "https://cloud.r-project.org")
BiocManager::install('SingleCellExperiment')
{%- endhighlight -%}

Of course, I have my actual SCE data given to me from my professor. Good look finding some on your own!

### Why is the SCE object?

Yes, I'm ignoring proper grammar for the joke. Something something RNA stuff.

### How is the SCE object?

Alright, I admit, it's getting old. But the structure of the Single-Cell Experiment is very important, and probably one of the most difficult things to explain.

<figure>
	<img src="{{site.url}}/{{site.baseurl}}/assets/img/scestructure.png" alt=""> 
	<figcaption>Figure 1. - Thanks to Bioconductor for their well-informed guide to SCE objects. Go check that out if you want a professional explanation. Stick around here if you want to watch an amateur make a fool of themselves.</figcaption>
</figure>

Oh, nevermind. It was pretty easy!

In all seriousness, an SCE object has an odd format. The primary assay, as mentioned before, is made of counts - it's essentially a matrix of mostly 0s that shows which cells express which genes. But then we have the metadata (imagine a dramatic sting here).

The metadata contains information about each cell. This is what happens when you load the column metadata on the first three cells.

`load("sce.Rdata")
first.3 <- sce[,1:3]
colData(first.3)`

{%- highlight python -%}
load("sce.Rdata")
first.3 <- sce[,1:3]
colData(first.3)
#=> prints DataFrame with 3 rows and 8 columns
                       nGene      nUMI      sample percent_mito doublet_score       batch        type       stage
                   <integer> <numeric> <character>    <numeric>     <numeric> <character> <character> <character>
AAACCTGAGAGCTTCT-9      1655      5721       S14_N 1.15962e-212    0.00133505           4           N      locAdv
AAACCTGAGTCGTTTG-9      1547      4883       S14_N  6.30512e-16    0.51172545           4           N      locAdv
AAACCTGGTGCATCTA-9      1349      2237       S14_N  5.01866e-15    0.00000000           4           N      locAdv
{%- endhighlight -%}

Okay what is this going to look like??

### Who is the -- okay, I'll stop now.
I know, I'm hilarious. ...this is probably why no one will hire me, huh.



## So now you want to analyze data.

Since I have to include some form of data analysis in this post, I suppose I will go over the basic things you can do with the SCE object.

### Normalization

Unfortunately, since this is biological data, it is very much not normal. In order to perform any kind of analysis, you're going to want to normalize it. 
(Sorry Python users - I'm working in R and I will not cater to your needs.)

{%- highlight python -%}
scaled <- t(t(assay(sce, "counts")) / colSums(assay(sce,"counts"))) # Takes counts assay and divide each table cell by the sum of the column counts to scale it.
logcounts <- log1p(scaled * 10000) # Create a matrix called 'logcounts' by multiplying the scaled matrix by 10000 and performing a log transformation.
assay(sce, "logcounts") <- logcounts # Inserts the new norm matrix as an assay in the SCE called logcounts.
{%- endhighlight -%}

The transposition stuff is just because R is whiny and doesn't like assays that are too tall. Me too, pal. The scaling factor isn't set in stone - I'm not sure why my professor chose 10,000. Ask him if you're curious.

Et voila (excuse my lack of an accent, I don't know how to type it in markdown), you have a normalized count assay!



### Code chunk!

{%- highlight python -%}
def print_hi(name):
  print("Hi" + name)
print_hi('Tom')
#=> prints 'Hi, Tom'.
{%- endhighlight -%}

Note to self: this is a tutorial blog post.

-install biocmanager
-install singlecellexperiment
